[Douglas Lim] iP

dlimyy opened the pull request:
# Duke
>Don’t Wait for Opportunity, Create it - George Bernard Shaw
## List of features 
* Add deadlines
* Add ToDos
* Add Events
* Remove tasks
* List tasks
* Find tasks

To get started, all you would need to do is,
1. Download it from [here](https://github.com/dlimyy/ip)
2. Go to releases
3. Download the jar file
4. Run the jar file
5. And now, all of your tasks can be organized!!! 🤩 

## Current Progress
- [X] Create tasks
- [X] Create GUI
- [ ] Beautify GUI

If you are a **java** programmer, you can use it to *practice* java too! Here is the `main` method:
```java
public class Main {
    public static void main() {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

dlimyy committed:
Level 1: Added the Greet, Echo and Exit features

dlimyy committed:
Level 2: Added Add and List features

dlimyy committed:
Removed trailing and leading whitespaces entered by user

dlimyy committed:
Level 3: Mark task as completed and uncompleted

dlimyy committed:
Level 4: Added ToDo, Event, Deadline tasks

dlimyy committed:
Added Automated Text UI Testing

dlimyy committed:
Added Exception Handling

dlimyy committed:
Updated toString method in DukeException

dlimyy committed:
Added Delete Feature

dlimyy committed:
Refactored Code

dlimyy committed:
Level 7: Added Save Feature

dlimyy committed:
Level 8: Add Date and time feature

Changed Deadline and Event tasks dates to LocalDate and handled case
where date passed in is invalid.

dlimyy committed:
Merge branch 'branch-Level-7'

Added Save feature

dlimyy committed:
Merge branch 'branch-Level-8'

Added Dates and Times for Deadline and Event tasks

dlimyy committed:
Use more OOP

Added Ui, Storage, Parser and various Command classes. Changed
unmark error message which had a typo.

dlimyy committed:
Organise the classes into the various packages

Added 3 packages, namely duke, duke.task and duke.command. The
classes were split into one of the 3 packages.

dlimyy committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Merged with gradle branch to add gradle files to master branch

dlimyy committed:
Change class name in Gradle File

Modified the main class name under application in Build.Gradle

dlimyy committed:
Add Junit tests

Added ParserTest which tests the Parser class.

dlimyy committed:
Add more JUnit test classes

Added EventTest and DeadlineTest classes.

dlimyy committed:
Changed Error message for WrongCommand

Changed WrongCommand message.

dlimyy committed:
Level 9: Add Find feature

Added FindCommand class to handle find command. Edited Parser class
to handle the FindCommand case. Edited Task class to check if the
task contains the keyword.

dlimyy committed:
Add JavaDoc to all of the classes and public methods

Added JavaDocs for all of the classes and to the public methods.

dlimyy committed:
Refactored code to meet coding standard

Changed from Integer class to int for position in deleteCommand.

dlimyy committed:
Merge branch 'branch-Level-9'

Add Find feature for the task.

dlimyy committed:
Add JavaDoc for FindCommand class

Added JavaDocs for the FindCommand class and methods.

dlimyy committed:
Merge branch 'branch-A-CodingStandard'

Refactored code and ensured that it adhered to coding standards.

tinenhao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+            } catch (DateTimeParseException e) {
+                ui.printBorder();
+                ui.showExceptionMessage("Sorry! Please include a valid date entry");
+                ui.printBorder();
```
I like how the catch blocks are differentiated with the corresponding error messages

tinenhao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+        }
+        if (command.toLowerCase().startsWith("find")) {
+            String keyword = parseFind(command);
+            return new FindCommand(keyword);
```
Consider using switch statement instead of if-else to make the code look cleaner

tinenhao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+package duke;
+
+import duke.command.*;
```
Should not import as wildcard

tinenhao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    public static Command parse(String command, TaskList taskList) throws DukeException,
+            DateTimeParseException, NumberFormatException{
+        if (command.toLowerCase().equals("bye")) {
```
Consider extracting out command.toLowerCase() as a String variable to reduce code duplication.

dlimyy committed:
Add Basic GUI

Added MainWindow, Main and DialogBox classes and their respective
FXML files. Added some padding to the label and resolved issue
where screen would shift slightly to the right.

dlimyy committed:
Merge branch 'branch-Level-10'

Added basic GUI.

lfrostbytee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    public String getReply(String input) {
+        Scanner scanner = new Scanner(System.in);
+        String reply;
```
I like how you have named your variable to be "reply". The naming helps me understand the bigger picture of how your code works.

lfrostbytee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+     * @throws IOException If the file cannot be found by scanner or I/O error occurred.
+     */
+    public ArrayList<Task> readFile() throws IOException {
+        ArrayList<Task> botArray  = new ArrayList<>();
```
Hmm... perhaps the name 'botArray' can be changed to something else more intuitive? The bot to me can seem like a "bottom array" or a "robot array".

lfrostbytee commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,41 @@
+ * The DeleteCommand class represents the delete command given by user.
+ */
+public class DeleteCommand extends Command {
+    private final int position;
```
I like the usage of the term "position" as it indicates that you are referring to the position (starting from index 1) instead of the common "index" (starting from index 0) which most programmers like to use. Using the position term clarifies the need of when you need to subtract by 1.

lfrostbytee commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Override
+    public String execute(Ui ui, Storage storage, TaskList taskList) {
+        MultiLineFormatter mf = new MultiLineFormatter();
```
I understand that the word "multilineformatter" can be too long to use as a variable name. Perhaps you could have the best of the both worlds such as naming it as "MFormatter"?

lfrostbytee commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,67 @@
+     * @param keyword Keyword that the task is being checked.
+     * @return true if the keyword is in the description;else return false;
+     */
+    public Boolean isContainsKeyword(String keyword) {
```
Perhaps a more intuitive name like: "hasKeyword" may be preferred?

lfrostbytee commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,28 @@
+public class DeadlineTest {
+
+    @Test
+    public void stringFormatting_unmarkedStringFormatting_success() {
```
Good job with following the naming style for Unit Testing!

lfrostbytee reviewed (commented):
LGTM! I had a hard time finding errors as the code quality was really good.

PokezardVGC commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+ * TaskList class which represents an array of tasks in the list.
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
if possible, make it private final

PokezardVGC commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    public String stringFormatting() {
+        String marker = getStatus().equals("X") ? "T" : "F";
+        return " # " + marker + " # " + this.description;
```
use String.format instead of concatination

dlimyy committed:
Fix some bugs in coding quality

Changed the import statment for Parser from duke.command.* to the various
duke.command classes. Renamed some variables and reduced code
duplication.

dlimyy committed:
Add Assertions into code

Added assert statements into Parser class.

dlimyy committed:
Merge pull request #1 from dlimyy/branch-A-Assertions

Add Assertions into code

dlimyy committed:
Improve code quality

Removed redundant uses of this in code. Changed some variable names.

dlimyy committed:
Merge pull request #2 from dlimyy/branch-A-CodeQuality

Improve code quality

dlimyy committed:
Add Undo Feature

Introduced Undo Feature and Added CommandHistory class.

dlimyy committed:
Add switch statments and abstract out code

Added switch statments and abstracted out some lines of code to
improve code quality.

dlimyy committed:
Update the user guide

Updated the markdown for the user guide.

dlimyy committed:
Add bye command to user guide

Added description and explanation of bye command in user guide.

dlimyy committed:
Add screenshot of the Duke Bot

Added screenshot of the Duke Bot product made.

dlimyy committed:
Improve code quality and update aesthetics of GUI

Improved code quality for all classes and changed the
colour of dialog boxes. Ui.png was also updated.

dlimyy committed:
Update formatting for ListCommand

Updated the formatting for message printed by
list command.

dlimyy committed:
Update user guide

Updated user guide by inserting horizontal lines
after each feature.

dlimyy committed:
Update README.md

Added Acknowledgements for images used.

dlimyy committed:
Update user guide

Added image to the user guide.

dlimyy committed:
Update build.gradle file

Update javafx version in build.gradle file to 11.0.2 to fix issue
where jar file is unable to run on Ubuntu systems.
